### PR TITLE
Discount for customer

### PR DESCRIPTION
**Pull request by boris-brtan**
- administration of customer now contains discount field with range validation that is reflected into User entity
- prices of products are recalculated for logged customers based on custom discount
- discount works also for companies with multiple users
- CurrentCustomer is also extended for discount needs

| Q             | A
| ------------- | ---
|Description, reason for the PR| we want to test the extensibility of pricing from Shopsys Framework
|New feature| Yes 
|BC breaks| No 
|Fixes issues| -
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: src/Shopsys/ShopBundle/Resources/translations/validators.en.po [hunk 1]
```diff
@@ -132,4 +141,3 @@ msgstr ""
 
 msgid "Zip code cannot be longer than {{ limit }} characters"
 msgstr ""
-
```
Comment (boris-brtan) [inline src/Shopsys/ShopBundle/Resources/translations/validators.en.po:135]: hmm this was strange

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
     - added new shipping method Zasilkovna
     - pick up places are downloaded by cron
 - [#15 - Company account with multiple users](https://github.com/shopsys/demoshop/pull/15)
+- [#31 - discount for customer](https://github.com/shopsys/demoshop/pull/31)
```
Comment (DavidOstrozlik) [inline CHANGELOG.md]: It's just a cosmetic issue. I also changed title of this PR.

```suggestion
- [#31 - Discount for customer](https://github.com/shopsys/demoshop/pull/31)
```

File: src/Shopsys/ShopBundle/Resources/translations/validators.cs.po [hunk 3]
```diff
@@ -10,6 +10,15 @@ msgstr "Název města nesmí být delší než {{ limit }} znaků"
...
 msgstr "Název firmy nesmí být delší než {{ limit }} znaků"
 
+msgid "Discount percentage should be {{ limit }} or more."
+msgstr "Procentní sleva musí být větší nebo rovna {{ limit }}"
+
+msgid "Discount percentage should be {{ limit }} or less."
+msgstr "Procentní sleva musí být menší nebo rovna {{ limit }}"
+
+msgid "Discount percentage needs to be valid number with range between 0 and 100"
+msgstr "Procentní Sleva musí být validní čísovka s rozmezím mezi 0 a 100"
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Resources/translations/validators.cs.po]:
```suggestion
msgstr "Procentní sleva musí být celé číslo v rozmezí 0 až 100"
```

**Review by DavidOstrozlik — CHANGES_REQUESTED** [review]
I reviewed this PR and found just 2 cosmetic changes.
Tests: I found that:
- you can't save User if you didn't fill the discount - fixed
- on Firefox you can fill discount with letters and it doesn't show the correct invalid message - problem for mozilla i will create issue https://github.com/shopsys/shopsys/issues/717
- on Firefox the JS validation of discount doesn't work - problem for mozilla i will create issue

**Comment by boris-brtan** [comment]
Hi, i did the fixes and created issue for the JSFormValidator for mozilla browsers

**Comment by boris-brtan** [comment]
I see the same thing as in promocode so i put 0 as default value and put there not blank constraints

**Comment by boris-brtan** [comment]
now there is need to test also filtering by price

File: src/Shopsys/ShopBundle/Model/Customer/User.php [hunk 4]
```diff
@@ -29,6 +29,12 @@ class User extends BaseUser
      */
     protected $billingAddress;
 
+    /**
+     * @var int|null
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/User.php]:
```suggestion
     * @var int
```

File: src/Shopsys/ShopBundle/Model/Customer/User.php [hunk 5]
```diff
@@ -39,6 +45,24 @@ public function __construct(
...
+     * @param \Shopsys\ShopBundle\Model\Customer\UserData $userData
+     */
+    public function edit(BaseUserData $userData)
+    {
+        $this->discount = $userData->discount;
+        parent::edit($userData);
+    }
+
+    /**
+     * @return int|null
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/User.php]:
```suggestion
     * @return int
```

File: src/Shopsys/ShopBundle/Model/Customer/User.php [hunk 6]
```diff
@@ -39,6 +45,24 @@ public function __construct(
...
+     */
+    public function getDiscount(): ?int
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/User.php]:
```suggestion
    public function getDiscount(): int
```

File: src/Shopsys/ShopBundle/Model/Customer/UserData.php [hunk 7]
```diff
@@ -10,4 +10,15 @@ class UserData extends BaseUserData
      * @var int|null
      */
     public $id;
+
+    /**
+     * @var int|null
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/UserData.php]:
```suggestion
     * @var int
```

File: src/Shopsys/ShopBundle/Model/Customer/CurrentCustomer.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+class CurrentCustomer extends BaseCurrentCustomer
+{
+    /**
+     * @return int|null
+     */
+    public function getCustomDiscount()
+    {
+        $user = $this->findCurrentUser();
+        if ($user === null) {
+            return null;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/CurrentCustomer.php]:
```suggestion
            return 0;
```

File: src/Shopsys/ShopBundle/Model/Customer/CurrentCustomer.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Shopsys\ShopBundle\Model\Customer;
+
+use Shopsys\FrameworkBundle\Model\Customer\CurrentCustomer as BaseCurrentCustomer;
+
+class CurrentCustomer extends BaseCurrentCustomer
+{
+    /**
+     * @return int|null
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Customer/CurrentCustomer.php]:
```suggestion
     * @return int
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        $this->productManualInputPriceRepository = $productManualInputPriceRepository;
+        $this->productRepository = $productRepository;
+        $this->pricingService = $pricingService;
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param int $domainId
+     * @param \Shopsys\FrameworkBundle\Model\Pricing\Group\PricingGroup $pricingGroup
+     * @param int|null $customDiscount
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
     * @param int $customDiscount
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+     * @return \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPrice
+     */
+    public function calculatePrice(Product $product, $domainId, PricingGroup $pricingGroup, $customDiscount = null)
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
    public function calculatePrice(Product $product, $domainId, PricingGroup $pricingGroup, $customDiscount)
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        }
+
+        return $this->calculateProductPriceForPricingGroup($product, $pricingGroup, $customDiscount);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $mainVariant
+     * @param int $domainId
+     * @param \Shopsys\FrameworkBundle\Model\Pricing\Group\PricingGroup $pricingGroup
+     * @param int|null $customDiscount
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
     * @param int $customDiscount
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+     * @return \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPrice
+     */
+    private function calculateMainVariantPrice(Product $mainVariant, $domainId, PricingGroup $pricingGroup, $customDiscount = null)
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
    private function calculateMainVariantPrice(Product $mainVariant, $domainId, PricingGroup $pricingGroup, $customDiscount)
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        $minVariantPrice = $this->pricingService->getMinimumPriceByPriceWithoutVat($variantPrices);
+        $from = $this->pricingService->arePricesDifferent($variantPrices);
+
+        return new ProductPrice($minVariantPrice, $from);
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param \Shopsys\FrameworkBundle\Model\Pricing\Group\PricingGroup $pricingGroup
+     * @param int|null $customDiscount
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
     * @param int $customDiscount
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+     * @return \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPrice
+     */
+    private function calculateProductPriceForPricingGroup(Product $product, PricingGroup $pricingGroup, $customDiscount = null)
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
    private function calculateProductPriceForPricingGroup(Product $product, PricingGroup $pricingGroup, $customDiscount)
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        $manualInputPrice = $this->productManualInputPriceRepository->findByProductAndPricingGroup($product, $pricingGroup);
+        if ($manualInputPrice !== null) {
+            $inputPrice = $manualInputPrice->getInputPrice() * (100 - $customDiscount ?? 100) / 100;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php]:
```suggestion
            $inputPrice = $manualInputPrice->getInputPrice() * (100 - $customDiscount) / 100;
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+class ProductPriceCalculationForUser extends BaseProductPriceCalculationForUser
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculation
+     */
+    private $productPriceCalculation;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Customer\CurrentCustomer
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php]:
```suggestion
     * @var \Shopsys\ShopBundle\Model\Customer\CurrentCustomer
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php [hunk 18]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Shopsys\ShopBundle\Model\Product\Pricing;
+
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Shopsys\FrameworkBundle\Model\Customer\CurrentCustomer;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php]:
```suggestion
use Shopsys\ShopBundle\Model\Customer\CurrentCustomer;
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    private $pricingGroupSettingFacade;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Domain\Domain
+     */
+    private $domain;
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculation $productPriceCalculation
+     * @param \Shopsys\FrameworkBundle\Model\Customer\CurrentCustomer $currentCustomer
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php]:
```suggestion
     * @param \Shopsys\ShopBundle\Model\Customer\CurrentCustomer $currentCustomer
```

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php [hunk 20]
```diff
@@ -0,0 +1,134 @@
+<?php
+
+namespace Shopsys\ShopBundle\Model\Product\Pricing;
+
+use Shopsys\FrameworkBundle\Model\Pricing\BasePriceCalculation;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculation.php:5]: Please change `FrameworkBundle` to `ShopBundle` in usages and annotations of classes that are extended in `ShopBundle`

File: src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace Shopsys\ShopBundle\Model\Product\Pricing;
+
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Shopsys\FrameworkBundle\Model\Customer\CurrentCustomer;
+use Shopsys\FrameworkBundle\Model\Customer\User;
+use Shopsys\FrameworkBundle\Model\Pricing\Group\PricingGroupSettingFacade;
+use Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculation;
+use Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculationForUser as BaseProductPriceCalculationForUser;
+use Shopsys\FrameworkBundle\Model\Product\Product;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Model/Product/Pricing/ProductPriceCalculationForUser.php:10]: Please change `FrameworkBundle` to `ShopBundle` in usages and annotations of classes that are extended in `ShopBundle`

File: src/Shopsys/ShopBundle/Controller/Front/ProductController.php [hunk 22]
```diff
@@ -134,6 +143,19 @@ public function detailAction($id)
         ]);
     }
 
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Filter\ProductFilterData $productFilterData
+     */
+    private function applyDiscountForPriceFilter(ProductFilterData $productFilterData)
+    {
+        if ($productFilterData->minimalPrice !== null) {
+            $productFilterData->minimalPrice /= ($this->currentCustomer->getCustomDiscount() ?? 100) / 100;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Controller/Front/ProductController.php]: This does not work properly

File: src/Shopsys/ShopBundle/Controller/Front/ProductController.php [hunk 23]
```diff
@@ -134,6 +143,19 @@ public function detailAction($id)
...
+        }
+        if ($productFilterData->maximalPrice !== null) {
+            $productFilterData->maximalPrice /= ($this->currentCustomer->getCustomDiscount() ?? 100) / 100;
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Controller/Front/ProductController.php]: this does not work properly

**Review by DavidOstrozlik — CHANGES_REQUESTED** [review]

File: src/Shopsys/ShopBundle/Controller/Front/ProductController.php [hunk 24]
```diff
@@ -224,6 +249,7 @@ public function listByBrandAction(Request $request, $id)
         $viewParameters = [
             'paginationResult' => $paginationResult,
             'brand' => $brand,
+            'discountCoeficient' => $this->currentCustomer->getDiscountCoeficient(),
```
Comment (DavidOstrozlik) [inline src/Shopsys/ShopBundle/Controller/Front/ProductController.php]: This parameter is not used anywhere
```suggestion
```

**Review by DavidOstrozlik — CHANGES_REQUESTED** [review]
I have just one suggestion and it would be great if you added information to "admin - order detail - add product" that the added product will not have any individual discounts

**Review by DavidOstrozlik — APPROVED** [review]

**Pull request merged**